### PR TITLE
Routes API

### PR DESCRIPTION
**Pull request by RoboSparrow**
**Proposal**, don't merge automatically!

Simple router API

**src/xAPI/Bootstrap.php**

* removed routes, they are now loaded from API/Routes and handed over to the Controller via a simple loop

**src/xAPI/Controller.php**

* simplified loader, removed name space compiling logic (performance)

**src/xAPI/Routes.php**

* provides and stores routes data.

**TODO**

Apply to extension API
* unit tests

### CODE REVIEW

**Comment by RoboSparrow** [comment]
I was also experimenting with loading the routes straight from the Config.yml. However, I refrained from it for this stage. Strategically it would be cool to have configurable (disable CORS, or even a read-only LRS)

File: src/xAPI/Bootstrap.php [hunk 1]
```diff
@@ -558,22 +567,16 @@ public function bootWebApp()
                     }
 
                     // Load any routes added by extension
-                    $routes = $extension->getRoutes();
-                    foreach ($routes as $route) {
-                        $app->map($route['methods'], $route['pattern'], [$extension, $route['callable']]);
-                    }
+                    $_routes = $extension->getRoutes();
```
Comment (sraka1) [inline src/xAPI/Bootstrap.php]: @RoboSparrow Why the _?

**Comment by RoboSparrow** [comment]
performance impact: it's faster, i dare saying that without benchmarking as the classmap is now static and not dynamic like the old one (controller::load __NAMESPACE__,$version, $resource, $subresource)

**Comment by RoboSparrow** [comment]
$_routes, can be replaced replace with $extensionRoutes

**Comment by sraka1** [comment]
$extensionRoutes is cool. 

Hm...you're right about performance. In future we need to have load routes by version (maybe by passing version to constructor), that was the reason that the mapping was dynamic there, so that when xAPI spec 1.1 comes out (with bigger changes), we can load from V11 namespace.

**Comment by RoboSparrow** [comment]
I have the extension com[ponent ready to push, just can't test due to auth issues

**Comment by RoboSparrow** [comment]
yeah,I had it in the constructor, in this case you would just fetch another array or merge with overwrite. tha'ts not a big deal. Th `V10` name-space is IMO not a good solution (i know it wasn’t your idea). We have `V10` namespaced a lot in `use Blah\V10\Bob as Blu;` headers. There are probably better approaches by manipulating the PSR4 autoloader etc.. 

```
    "autoload": {
        "psr-4": {
            "API\\": "src/xAPI"
            "APIv10\": "src/v10"
            "APIv11": "src/v11"
        }
}
```

**Comment by sraka1** [comment]
It would've been if xAPI would change more frequently, but now that it looks like 1.1 won't be here for  a while....it looks like a redundant layer.

**Pull request merged**